pkg/minikube/bootstrapper/kubeadm/kubeadm.go

```diff
@@ -0,0 +1,354 @@
+	}
+	return "", fmt.Errorf("Error: Unrecognized output from ClusterStatus: %s", status)
+}
+
+// TODO(r2d4): Should this aggregate all the logs from the control plane?
+// Maybe subcommands for each component? minikube logs apiserver?
+func (k *KubeadmBootstrapper) GetClusterLogs(follow bool) (string, error) {
+	var flags []string
+	if follow {
+		flags = append(flags, "-f")
+	}
+	logsCommand := fmt.Sprintf("sudo journalctl %s -u kubelet", strings.Join(flags, " "))
+
+	if follow {
+		if err := k.c.Run(logsCommand); err != nil {
+			return "", errors.Wrap(err, "getting shell")
+		}
+	}
+
+	logs, err := k.c.CombinedOutput(logsCommand)
+	if err != nil {
+		return "", errors.Wrap(err, "getting cluster logs")
+	}
+
+	return logs, nil
+}
+
+func (k *KubeadmBootstrapper) StartCluster(k8s bootstrapper.KubernetesConfig) error {
+	// We use --skip-preflight-checks since we have our own custom addons
+	// that we also stick in /etc/kubernetes/manifests
+	kubeadmTmpl := "sudo /usr/bin/kubeadm init --config {{.KubeadmConfigFile}} --skip-preflight-checks"
+	t := template.Must(template.New("kubeadmTmpl").Parse(kubeadmTmpl))
+	b := bytes.Buffer{}
+	if err := t.Execute(&b, struct{ KubeadmConfigFile string }{constants.KubeadmConfigFile}); err != nil {
+		return err
+	}
+
+	err := k.c.Run(b.String())
+	if err != nil {
+		return errors.Wrapf(err, "kubeadm init error running command: %s", b.String())
+	}
+
+	//TODO(r2d4): get rid of global here
+	master = k8s.NodeName
+	if err := util.RetryAfter(100, unmarkMaster, time.Millisecond*500); err != nil {
+		return errors.Wrap(err, "timed out waiting to unmark master")
+	}
+
+	if err := util.RetryAfter(100, elevateKubeSystemPrivileges, time.Millisecond*500); err != nil {
+		return errors.Wrap(err, "timed out waiting to elevate kube-system RBAC privileges")
+	}
+
+	return nil
+}
+
+//TODO(r2d4): Split out into shared function between localkube and kubeadm
+func addAddons(files *[]assets.CopyableFile) error {
+	// add addons to file list
+	// custom addons
+	assets.AddMinikubeDirToAssets("addons", constants.AddonsPath, files)
+	// bundled addons
+	for addonName, addonBundle := range assets.Addons {
+		// TODO(r2d4): Kubeadm ignores the kube-dns addon and uses its own.
+		// expose this in a better way
+		if addonName == "kube-dns" {
+			continue
+		}
+		if isEnabled, err := addonBundle.IsEnabled(); err == nil && isEnabled {
+			for _, addon := range addonBundle.Assets {
+				*files = append(*files, addon)
+			}
+		} else if err != nil {
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func (k *KubeadmBootstrapper) RestartCluster(k8s bootstrapper.KubernetesConfig) error {
+	restoreTmpl := `
+	sudo kubeadm alpha phase certs all --config {{.KubeadmConfigFile}} &&
+	sudo /usr/bin/kubeadm alpha phase kubeconfig all --config {{.KubeadmConfigFile}} &&
+	sudo /usr/bin/kubeadm alpha phase controlplane all --config {{.KubeadmConfigFile}} &&
+	sudo /usr/bin/kubeadm alpha phase etcd local --config {{.KubeadmConfigFile}}
+	`
```

Comment:
each line in this string are going to be indented. is that ok?

Reply:
I'll undent it. I don't think it matters either way